Data/ByteString/Lazy/Char8.hs

```diff
@@ -974,40 +1111,41 @@ readInt bs = case L.uncons bs of
             BI.accursedUnutterablePerformIO $
                 BI.unsafeWithForeignPtr fp $ \ptr -> do
                     let end = ptr `plusPtr` len
-                    x@(!_, !_, !_) <- if positive
-                        then digits intmaxQuot10 intmaxRem10 end ptr 0 acc
-                        else digits intminQuot10 intminRem10 end ptr 0 acc
+                    x@(!_, !_, !_) <- digits end ptr 0 acc
                     return x
           where
-            digits !maxq !maxr !e !ptr = go ptr
+            digits !e !ptr = go ptr
               where
-                go :: Ptr Word8 -> Int -> Word -> IO (Int, Word, Bool)
+                go :: Ptr Word8 -> Int -> Word64 -> IO (Int, Word64, Bool)
                 go !p !b !a | p == e = return (b, a, True)
                 go !p !b !a = do
-                    !w <- fromIntegral <$> peek p
+                    !w <- fromIntegral @Word8 @Word64 <$> peek p
                     let !d = w - 0x30
                     if d > 9 -- No more digits
                         then return (b, a, True)
-                        else if a < maxq -- Look for more
+                        else if a < word64maxQuot10 -- Look for more
                         then go (p `plusPtr` 1) (b + 1) (a * 10 + d)
-                        else if a > maxq -- overflow
+                        else if a > word64maxQuot10 -- overflow
                         then return (b, a, False)
-                        else if d <= maxr -- Ideally this will be the last digit
+                        else if d <= word64maxRem10 -- Ideally the last digit
                         then go (p `plusPtr` 1) (b + 1) (a * 10 + d)
                         else return (b, a, False) -- overflow
 
         -- | Plausible success, provided we got at least one digit!
-        result !nbytes !acc str
-            | nbytes > 0 = let !i = w2int acc in Just (i, str)
+        result !nbytes !w str
+            | nbytes > 0 = Just (w, str)
             | otherwise  = Nothing
 
-        -- This assumes that @negate . fromIntegral@ correctly produces
-        -- @minBound :: Int@ when given its positive 'Word' value as an
-        -- input.  This is true in both 2s-complement and 1s-complement
-        -- arithmetic, so seems like a safe bet.  Tests cover this case,
-        -- though the CI may not run on sufficiently exotic CPUs.
-        w2int !n | positive = fromIntegral n
-                 | otherwise = negate $! fromIntegral n
+-- | Bounds for Word64 multiplication by 10 without overflow, and
+-- Word64 values of Word and Int bounds.
+word64maxQuot10, word64maxRem10, wordMaxAsWord64 :: Word64
+word64maxQuot10 = maxBound `div` 10
+word64maxRem10  = maxBound `rem` 10
+wordMaxAsWord64 = fromIntegral @Word @Word64 maxBound
+
+intMaxAsWord64, intMinAsWord64 :: Word64
+intMaxAsWord64  = fromIntegral @Int @Word64 maxBound
+intMinAsWord64  = fromIntegral @Word @Word64 $ fromIntegral @Int @Word minBound
```

Comment:
Is this wrong?

```suggestion
intMinAsWord64  = fromIntegral @Int @Word64 minBound
```

Two's complement is not my forte…

Reply:
No, the original is correct, and the shorter form gives the wrong answer on 32-bit systems.